### PR TITLE
ML-DSA 44 instantiations and multiplexing

### PR DESCRIPTION
**Pull request by franziskuskiefer**
- add build.rs for feature detection
- add instantiations for 44
- add neon sha3

Towards #275

### CODE REVIEW

**Review by keks — COMMENTED** [review]
So far I only have pretty high level comments, as this code is pretty dense with meta programming and I always find it difficult to get into. I do have two style questions. It seems like the tests are still failing, so maybe this isn't quite ready yet anyway?

File: libcrux-ml-dsa/src/ml_dsa_44.rs [hunk 1]
```diff
@@ -60,39 +65,125 @@ pub type MLDSA44VerificationKey = MLDSAVerificationKey<VERIFICATION_KEY_SIZE>;
...
+// #[cfg(feature = "simd256")]
+// type Shake256X4 = crate::hash_functions::simd256::Shake256X4;
+// #[cfg(not(feature = "simd256"))]
+// type Shake256X4 = crate::hash_functions::portable::Shake256X4;
+
+// // TODO: This is all portable for now.
+// #[cfg(feature = "simd256")]
+// type Shake256 = crate::hash_functions::portable::Shake256;
+// #[cfg(not(feature = "simd256"))]
+// type Shake256 = crate::hash_functions::portable::Shake256;
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply libcrux-ml-dsa/src/ml_dsa_44.rs:91]: No ... I'll remove it.

File: libcrux-ml-dsa/Cargo.toml [hunk 2]
```diff
@@ -15,14 +15,15 @@ publish = false
...
+libcrux-platform = { version = "0.0.2-alpha.3", path = "../sys/platform" }
 
 [dev-dependencies]
 rand = { version = "0.8" }
 hex = { version = "0.4.3", features = ["serde"] }
 serde_json = { version = "1.0" }
 serde = { version = "1.0", features = ["derive"] }
 criterion = "0.5"
-pqcrypto-dilithium = { version = "0.5.0" } #, default-features = false
+pqcrypto-dilithium = { version = "0.5.0" }         #, default-features = false
```
[the remark this answers is not in the document]
Comment (franziskuskiefer) [reply libcrux-ml-dsa/Cargo.toml:26]: auto formatting ... And, yes I think it should be there :)

**Review by franziskuskiefer — COMMENTED** [review]
Yeah, no one else is around right now to review and I thought it could be nice to get into 😉 You don't need to go too deep.

> It seems like the tests are still failing

Github seems to have issues.

File: libcrux-ml-dsa/build.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    let enable_simd128 = read_env("LIBCRUX_ENABLE_SIMD128");
+    let enable_simd256 = read_env("LIBCRUX_ENABLE_SIMD256");
+
+    let simd128_possible = target_arch == "aarch64";
+    if (simd128_possible || enable_simd128) && !disable_simd128 {
+        // We enable simd128 on all aarch64 builds.
+        println!("cargo:rustc-cfg=feature=\"simd128\"");
+    }
+    let simd126_possible = target_arch == "x86_64";
+    if (simd126_possible || enable_simd256) && !disable_simd256 {
```
Comment (jschneider-bensch) [inline libcrux-ml-dsa/build.rs:18]:
```suggestion
    let simd256_possible = target_arch == "x86_64";
    if (simd256_possible || enable_simd256) && !disable_simd256 {
```

File: libcrux-ml-dsa/build.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    let disable_simd256 = read_env("LIBCRUX_DISABLE_SIMD256");
+
+    // Force a simd build. Make sure you know what you're doing.
+    let enable_simd128 = read_env("LIBCRUX_ENABLE_SIMD128");
+    let enable_simd256 = read_env("LIBCRUX_ENABLE_SIMD256");
+
+    let simd128_possible = target_arch == "aarch64";
+    if (simd128_possible || enable_simd128) && !disable_simd128 {
+        // We enable simd128 on all aarch64 builds.
+        println!("cargo:rustc-cfg=feature=\"simd128\"");
```
Comment (jschneider-bensch) [inline libcrux-ml-dsa/build.rs:15]: I thought we would need `"cargo::rustc-cfg=feature=\"simd128\""` with double `:`, but I guess it also works like this?

File: libcrux-ml-dsa/src/hash_functions.rs [hunk 5]
```diff
@@ -480,60 +440,32 @@ pub(crate) mod simd256 {
 pub(crate) mod neon {
     // FIXME: This is only a portable implementation for now.
```
Comment (jschneider-bensch) [inline libcrux-ml-dsa/src/hash_functions.rs:441]:
```suggestion
```

**Review by jschneider-bensch — APPROVED** [review]
Looks good to me! Just some incidental things below.

File: libcrux-ml-dsa/build.rs (hunk 4 above)
Comment (franziskuskiefer) [reply libcrux-ml-dsa/build.rs:15]: `::` is the new version. But the old things keep working. We still had some older compilers and issue with `:`. I guess we could move to `::` soonish. But we should make sure to set the MSRV (1.77) accordingly when we do.

**Pull request merged**